lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -1298,6 +1298,39 @@ HRESULT ValidateLoadModuleFromContainerLazy(
                                          /*bLazyLoad*/ true);
 }
 
+HRESULT ValidateDxilContainerAgainstModule(const void *pContainer,
+                                           uint32_t ContainerSize,
+                                           llvm::Module *pModule,
+                                           llvm::Module *pDebugModule,
+                                           llvm::raw_ostream &DiagStream) {
+  const DxilPartHeader *pPart = nullptr;
+  IFR(FindDxilPart(pContainer, ContainerSize, DFCC_DXIL, &pPart));
+  // Verify that the ProgramHeader in the container contains a program version
+  // that matches the DxilModule's shader model version
+  const DxilProgramHeader *pProgramHeader =
+      reinterpret_cast<const DxilProgramHeader *>(GetDxilPartData(pPart));
+  if (pProgramHeader) {
+    int PV = pProgramHeader->ProgramVersion;
+    int major = (PV >> 4) & 0xF; // Extract the major version (next 4 bits)
+    int minor = PV & 0xF;        // Extract the minor version (lowest 4 bits)
+    DxilModule *pDxilModule = DxilModule::TryGetDxilModule(pModule);
+    if (!pDxilModule) {
+      return DXC_E_IR_VERIFICATION_FAILED;
+    }
+    ValidationContext ValCtx(*pModule, pDebugModule, *pDxilModule);
+    int moduleMajor = ValCtx.DxilMod.GetShaderModel()->GetMajor();
+    int moduleMinor = ValCtx.DxilMod.GetShaderModel()->GetMinor();
+    if (moduleMajor != major || moduleMinor != minor) {
```

Comment:
I think this validation would be better suited under `ValidateDxilContainerParts`.  This check can simply be triggered under both of these cases that are currently skipped:

https://github.com/microsoft/DirectXShaderCompiler/blob/25faa88053437c30e599b9e5323c94addc2397a1/lib/DxilValidation/DxilContainerValidation.cpp#L1036-L1037

You'll already have the `pPart`, `ValCtx`, and `pDxilModule`.

Plus, you can `GetShaderModel()` directly from `pDxilModule` instead of going through `ValCtx.DxilMod`.